src/gui/src/drcWidget.cpp

```diff
@@ -0,0 +1,538 @@
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include <QDebug>
+
+#include <QHeaderView>
+#include <QVBoxLayout>
+#include <QFileDialog>
+
+#include <map>
+#include <fstream>
+#include <regex>
+#include <sstream>
+#include <iomanip>
+
+#include "drcWidget.h"
+#include "utl/Logger.h"
+
+Q_DECLARE_METATYPE(gui::DRCViolation*);
+
+namespace gui {
+
+static QString getBBoxString(const odb::Rect& bbox, odb::dbBlock* block)
+{
+  double to_microns = block->getDbUnitsPerMicron();
+  std::stringstream ss;
+  ss << std::fixed << std::setprecision(5) << "(";
+  ss << bbox.xMin() / to_microns << ",";
+  ss << bbox.yMin() / to_microns << "), (";
+  ss << bbox.xMax() / to_microns << ",";
+  ss << bbox.yMax() / to_microns << ")";
+  return QString::fromStdString(ss.str());
+}
+
+///////
+
+DRCViolation::DRCViolation(const std::string& type,
+                           const std::vector<std::any>& srcs,
+                           const std::vector<DRCShape>& shapes,
+                           odb::dbTechLayer* layer,
+                           const std::string& comment) :
+                               type_(type),
+                               srcs_(srcs),
+                               shapes_(shapes),
+                               layer_(layer),
+                               comment_(comment)
+{
+  computeBBox();
+}
+
+DRCViolation::DRCViolation(const std::string& type,
+                           const std::vector<DRCShape>& shapes,
+                           const std::string& comment) :
+                                DRCViolation(type,
+                                             {},
+                                             shapes,
+                                             nullptr,
+                                             comment) {}
+
+void DRCViolation::computeBBox()
+{
+  QPolygon outline;
+  for (const auto& shape : shapes_) {
+    if(auto s = std::get_if<QLine>(&shape)) {
+      outline << (*s).p1() << (*s).p2();
+    } else if (auto s = std::get_if<QRect>(&shape)) {
+      outline = outline.united(*s);
+    } else if (auto s = std::get_if<QPolygon>(&shape)) {
+      outline = outline.united(*s);
+    }
+  }
+
+  QRect bounds = outline.boundingRect();
+  bbox_ = odb::Rect(
+      bounds.left(),
+      bounds.bottom(),
+      bounds.right(),
+      bounds.top());
+}
+
+void DRCViolation::paint(Painter& painter)
+{
+  for (const auto& shape : shapes_) {
+    if(auto s = std::get_if<QLine>(&shape)) {
+      const QPoint& p1 = (*s).p1();
+      const QPoint& p2 = (*s).p2();
+      painter.drawLine(p1.x(), p1.y(), p2.x(), p2.y());
+    } else if (auto s = std::get_if<QRect>(&shape)) {
+      const QRect rect = *s;
+      painter.drawRect({rect.left(), rect.bottom(), rect.right(), rect.top()});
+    } else if (auto s = std::get_if<QPolygon>(&shape)) {
+      // TODO: add polygon
```

Comment:
You might look at Graphics::drawPolygon90Set in fin where I used Boost to decompose the polygon to rects.